### ISSUE TITLE
re-export `from typing import Literal as L`: F821 Undefined name `bar`

### ISSUE THREAD

**Issue by Jasha10** [astral-sh/ruff#6014]
Hello,

If I import `typing.Literal` in `lib.py` then do `from lib import Literal` in `main.py`, ruff fails to treat the `Literal` type annotations properly.

```python
# lib.py
from typing import Literal

__all__ = ["Literal"]
```
```python
# main.py
from lib import Literal

foo: Literal["bar"] = "bar"
```
```bash
$ ruff check *.py
main.py:4:15: F821 Undefined name `bar`
Found 1 error.
$ ruff --version
ruff 0.0.280
```

### Expected behavior:

Ruff should recognize that `Literal["bar"]` is an annotation for a string literal and not flag the name `bar` as undefined.

**Comment by charliermarsh** [comment]
For now, you'll need to mark `lib` in [`typing-modules`](https://beta.ruff.rs/docs/settings/#typing-modules), like:

```python
[tool.ruff]
typing-modules = ["lib"]
```

This tells Ruff to treat exports from `lib` as if they are re-exports from `typing`. Let me know if it doesn't work for whatever reason :)

**Comment by Jasha10** [comment]
Thanks for the tip, @charliermarsh.
This still does not work for my use-case as, in `lib.py`, I have `from typing import Literal as L`. Here's an updated repro:

```python
# lib.py
from typing import Literal as L
from typing import Literal

__all__ = ["L", "Literal"]
```
```python
# main.py
from lib import L, Literal

foo1: Literal["bar"] = "bar"
foo2: L["bar"] = "bar"
```
```toml
# pyproject.toml
[tool.ruff]
typing-modules = ["lib"]
```
```bash
$ ruff check *.py
main.py:5:10: F821 Undefined name `bar`
Found 1 error.
```

I'll update this issue's name accordingly and re-open the issue.

**Comment by charliermarsh** [comment]
Makes sense, thanks. To set expectations, this won't be fixed in an imminent release since it requires us to resolve symbols across files, but it will be fixed eventually.

**Comment by rlipperts** [comment]
I also stumbled upon this while evaluating if I want to use [nptyping](https://github.com/ramonhagenaars/nptyping) in my work project

```python
from nptyping import Float, NDArray, Shape

my_array: NDArray[Shape["x, y, z, 3"], Float]
```

results in

```bash
$ ruff check nptyping_test.py
nptyping_test.py:3:26: F821 Undefined name `x`
nptyping_test.py:3:29: F821 Undefined name `y`
nptyping_test.py:3:32: F821 Undefined name `z`
Found 3 errors.
```

even with 

```toml
typing-modules = ["nptyping"]
```

in my `pyproject.toml`. 

As it seems this is a real-world example of the issue described by @Jasha10 - [nptyping uses Literal for their shape Keyword](https://github.com/ramonhagenaars/nptyping/blame/master/USERDOCS.md#L118C00-L127C60).

**Comment by Jasha10** [comment]
Thanks for adding this context @rlipperts. My usecase is also related to `ntyping`, as I'm using a personal fork of `ntyping` in my own project

**Comment by charliermarsh** [comment]
We could special-case `nptying.Shape` as an alias of `Literal`. (But it wouldn't solve cases in which that's re-exported under some other alias.)

**Comment by rlipperts** [comment]
As outline in the referenced link it is possible to replace `Shape` with `Literal`. For me that is Hotfix enough so that there is no need to clutter your codebase, @charliermarsh,  if you intend to fix it properly in the long run.

**Comment by Pierre-Sassoulas** [comment]
We also encounter something very similar in pylint when upgrading to 0.0.280, but on the actual ``typing.Litteral`` and not an alias:

```
pylint/utils/utils.py:56:31: F821 Undefined name `ignored`
pylint/utils/utils.py:56:39: F821 Undefined name `modules`
pylint/utils/utils.py:64:36: F821 Undefined name `py`
pylint/utils/utils.py:64:39: F821 Undefined name `version`
Found 4 errors.
```

When there are multiple values in the ``Literal`` it seems to works:
Code:

https://github.com/pylint-dev/pylint/blob/a57dd01c465a1357fc96833bec278fbf47147a4c/pylint/utils/utils.py#L27C1-L64

**Comment by charliermarsh** [comment]
@Pierre-Sassoulas - I believe that issue is separate and related to the use of future annotations. It’s fixed on main and will be out soon. (Similar to https://github.com/astral-sh/ruff/issues/6055.)
